HW2/P3/mandelbrot.pyx

```diff
@@ -4,68 +4,114 @@ cimport cython
 import numpy
 cimport AVX
 from cython.parallel import prange
-
-
-cdef np.float64_t magnitude_squared(np.complex64_t z) nogil:
-    return z.real * z.real + z.imag * z.imag
-
+        
 @cython.boundscheck(False)
 @cython.wraparound(False)
+        
 cpdef mandelbrot(np.complex64_t [:, :] in_coords,
                  np.uint32_t [:, :] out_counts,
                  int max_iterations=511):
     cdef:
-       int i, j, iter
-       np.complex64_t c, z
-
-       # To declare AVX.float8 variables, use:
-       # cdef:
-       #     AVX.float8 v1, v2, v3
-       #
-       # And then, for example, to multiply them
-       #     v3 = AVX.mul(v1, v2)
-       #
-       # You may find the numpy.real() and numpy.imag() fuctions helpful.
-
+        int i, j, iter
+        
+        #wrote each variable individually to make sure none left out      
+        AVX.float8 z_mag
+        AVX.float8 z_real
+        AVX.float8 z_imag 
+        AVX.float8 z_real_new 
+        AVX.float8 z_imag_new 
+        AVX.float8 cr
+        AVX.float8 ci 
+        AVX.float8 mask 
+        AVX.float8 const_one 
+        AVX.float8 const_four
+        AVX.float8 counts 
+        AVX.float8 const_zero 
+        
+        float [:,:] real, imag
+    
+    #grab the real and im. parts of in_coords before we give up gil
+    real = np.real(in_coords)
+    imag = np.imag(in_coords)
+    
     assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
     assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
-    assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
+    assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"  
 
     with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
+        
+        #Parallelize over 1,2, and 4 threads. 4 threads shown here
+        for i in prange(in_coords.shape[0],num_threads=4,schedule='static',chunksize=1):
+            for j in range(in_coords.shape[1]/8): #note: divide by 8 since we will grab 8 elements at a time 
+                
+                #create float8 constants to be used in operations below
+                const_zero = AVX.float_to_float8(0.0)
+                const_one = AVX.float_to_float8(1.0) 
+                const_four = AVX.float_to_float8(4.0) 
```

Comment:
Nice! In retrospect, a much better naming strategy than mine (counts, counts_add, and exit_check)! :)